### PR TITLE
05_debug: solution inplementation

### PR DESCRIPTION
**Pull request by ASynenko**
Module with sysfs and procfs interface for string case convertion.

### CODE REVIEW

File: 05_debug/debug_module.c [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+	while (conv->str_buff[indx] != '\0' &&
+		conv->str_buff[indx] != '\n' &&
+		indx < str_size) {
+		/*
+		 *!( isupper XOR to_lover )
+		 * |   t      |    t      | t - lovercase output
+		 * |   f      |    f      | t - uppercase output
+		 */
+		if (!(isupper(conv->str_buff[indx]) ^ to_lover)) {
```
Comment (ReAlUA) [inline 05_debug/debug_module.c]: At first, `isupper()` returns **int** value which can be, say, `0x20`.  Common `ctype.h` functions' implementation contains 256-byte array with bit masks that describes characters and `is`-function just returns a value from the array masked with appropriate mask (`isalnum` uses an OR of UPPER, LOWER and DIGIT masks). You should not relay on actual UPPER mask even it is 0x01 in the current implementation.
Use an idiom`!!` to convert to bool.

But... As far as I remember, `toupper` functions guarantees that it return unchanged argument if it isn't a lowercase character, the same for `tolower`. This `if` is redundant. Perhups, this code will be more compact and fast. Perhups not :-)
```c
    char ch = to_lower ? tolower(...) : toupper(...);
    if (ch != conv->str_buff[indx])
        conv->converted_last++;
    conv->str_buff[indx] = ch;
```

File: 05_debug/debug_module.c [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+	}
+	conv->processed_last  = indx;
+
+	conv->converted_total += conv->converted_last;
+	conv->processed_total += conv->processed_last;
+
+	return 0;
+}
+
+int print_stat(char *buf, struct convertor_struct *conv)
```
Comment (ReAlUA) [inline 05_debug/debug_module.c]: Make `static`

File: 05_debug/debug_module.c [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+{
+	if (conv == NULL || buf == NULL) {
+		pr_err("%s: convertor or buffer is NULL\n", __func__);
+		return -EINVAL;
+	}
+
+	return sprintf(buf, "processed last : %d char(s)\n"
+			"converted last : %d char(s)\n"
+			"converted total: %d char(s)\n"
+			"processed total: %d char(s)\n",
```
Comment (ReAlUA) [inline 05_debug/debug_module.c]: Different order of processed/converted for last/total is inconvenient to read.

File: 05_debug/debug_module.c [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+}
+
+static int debug_module_init(void)
+{
+	int ret;
+
+	pr_info("%s: module starting\n",  __func__);
+
+	kzfree(procfs_conv.str_buff);
+	kzfree(sysfs_conv.str_buff);
```
Comment (ReAlUA) [inline 05_debug/debug_module.c]: Wrong place. Did you want to place them into exit function?

File: 05_debug/debug_module.c [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+
+struct convertor_struct {
+	char *str_buff;
+
+	int converted_last;
+	int converted_total;
+	int processed_last;
+	int processed_total;
+
+} sysfs_conv, procfs_conv;
```
Comment (ReAlUA) [inline 05_debug/debug_module.c]: Use separate type declaration and two `static struct ...;`

**Review by ReAlUA — CHANGES_REQUESTED** [review]

File: 05_debug/debug_module.c (hunk 4 above)
Comment (ASynenko) [reply 05_debug/debug_module.c]: Yes. My mistake, will fix

File: 05_debug/debug_module.c (hunk 1 above)
Comment (ASynenko) [reply 05_debug/debug_module.c]: Changed to suggested solution

File: 05_debug/debug_module.c (hunk 2 above)
Comment (ASynenko) [reply 05_debug/debug_module.c]: Fixed

File: 05_debug/debug_module.c (hunk 3 above)
Comment (ASynenko) [reply 05_debug/debug_module.c]: Fixed

File: 05_debug/debug_module.c (hunk 5 above)
Comment (ASynenko) [reply 05_debug/debug_module.c]: Fixed

Comment (ReAlUA) [reply 05_debug/debug_module.c]: Remove ` sysfs_conv, procfs_conv` above and add `static` below.

File: 05_debug/debug_module.c [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+	if (conv == NULL || conv->str_buff == NULL) {
+		pr_err("%s: convertor or buffer is NULL\n", __func__);
+		return -EINVAL;
+	}
+
+	while (conv->str_buff[indx] != '\0' &&
+		conv->str_buff[indx] != '\n' &&
+		indx < str_size) {
+			char ch = to_lover ? tolower(conv->str_buff[indx]) : toupper(conv->str_buff[indx]);
+			if (ch != conv->str_buff[indx]) {
```
Comment (ReAlUA) [inline 05_debug/debug_module.c]: Add an empty line between variable drclaration and `if` statement.

File: 05_debug/debug_module.c [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+	size_t indx = 0;
+
+	conv->converted_last = 0;
+
+	if (conv == NULL || conv->str_buff == NULL) {
+		pr_err("%s: convertor or buffer is NULL\n", __func__);
+		return -EINVAL;
+	}
+
+	while (conv->str_buff[indx] != '\0' &&
```
Comment (ReAlUA) [inline 05_debug/debug_module.c:54]: Fix indentation of `while` body.
* Condition continuation -- 2 TAB
* Body -- 1 TAB

Fix indentation of `if` body (for now there is different indentation of body lines.

**Review by ReAlUA — CHANGES_REQUESTED** [review]
Everything work fine, but some cosmetic changes are needed.

File: 05_debug/debug_module.c (hunk 5 above)
Comment (ASynenko) [reply 05_debug/debug_module.c]: Fixed

File: 05_debug/debug_module.c (hunk 6 above)
Comment (ASynenko) [reply 05_debug/debug_module.c]: Fixed

File: 05_debug/debug_module.c (hunk 7 above)
Comment (ASynenko) [reply 05_debug/debug_module.c:54]: Fixed

Comment (ReAlUA) [reply 05_debug/debug_module.c:54]: For now, indentation of the inner `if` block is two TAB from the outer `while` level but must be one.

File: 05_debug/debug_module.c [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+		conv->str_buff[indx] != '\n' &&
```
Comment (ReAlUA) [inline 05_debug/debug_module.c]: Remove `'\n'` check to enable a multi-line string conversion (e.g. file).

**Review by ReAlUA — CHANGES_REQUESTED** [review]

**Review by ReAlUA — APPROVED** [review]

**Pull request merged**